File: internals/distcheck/src/distcheck/node_flow/platforms.py
"""Host-platform detection for the node distcheck flow (#520).

Reads the published-target table from `packages/ts/src/platforms.json`, the one
declarative copy shared with the TypeScript SDK, and keeps the fields this flow
needs to reconstruct the host's `@dirsql/lib-<slug>` sub-package (name, os/cpu
constraints, addon filename) and locate its staged addon. The lookup is pure --
callers pass the detected `sys.platform` / `platform.machine()`.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from .arch import to_node_arch

_LIB_PREFIX = "@dirsql/lib-"
_TABLE = Path(__file__).parents[5] / "packages" / "ts" / "src" / "platforms.json"


@dataclass(frozen=True)
class Platform:
    node_platform: str
    node_arch: str
    slug: str
    os: list[str]
    cpu: list[str]

    @property
    def name(self) -> str:
        """napi addon sub-package name (`@dirsql/lib-<slug>`).

        Since #739 this is the only per-platform family: the addon carries
        the CLI (`runCli`), so no `@dirsql/cli-<slug>` is published.
        """
        return f"@dirsql/lib-{self.slug}"

    @property
    def addon_name(self) -> str:
        """The `.node` filename napi emits for this triple."""
        return f"dirsql.{self.slug}.node"


PLATFORMS: tuple[Platform, ...] = tuple(
    Platform(
        row["nodePlatform"],
        row["nodeArch"],
        row["libName"].removeprefix(_LIB_PREFIX),
        row["os"],
        row["cpu"],
    )
    for row in json.loads(_TABLE.read_text(encoding="utf-8"))
)

_BY_KEY = {(p.node_platform, p.node_arch): p for p in PLATFORMS}


def find_host_platform(node_platform: str, node_arch: str) -> Platform:
    platform = _BY_KEY.get((node_platform, node_arch))
    if platform is None:
        raise ValueError(
            f"unsupported host {node_platform}-{node_arch}; "
            "add a row to packages/ts/src/platforms.json"
        )
    return platform


def detect_host(sys_platform: str, machine: str) -> Platform:
    return find_host_platform(sys_platform, to_node_arch(machine))

File: packages/ts/src/platforms.ts
// The target platforms `dirsql` publishes. The rows live in `platforms.json`,
// the one declarative copy of the table: this module types it, and the node
// distcheck flow (`internals/distcheck/.../node_flow/platforms.py`) reads the
// same file. The JSON sits under `src/` so `tsc` emits it into `dist/` and it
// ships with the package -- `loadNativeCore()` needs the map at runtime.
//
// Every target triple generates ONE npm sub-package: `@dirsql/lib-<slug>`,
// holding the napi-rs `.node` addon. It backs both layers -- the TypeScript
// SDK loads it via `loadNativeCore()`, and since #739 the `dirsql` CLI runs
// in-process through its `runCli` export rather than spawning a binary. The
// second family, `@dirsql/cli-<slug>`, shipped a redundant copy of the same
// core and is gone.
//
// The sub-packages are `optionalDependencies` of the main `dirsql` package,
// so npm/pnpm install only the one matching the host's OS/arch.
//
// `libTriples()` returns a `${process.platform}-${process.arch}` →
// sub-package-name map.

import rows from "./platforms.json" with { type: "json" };

export interface Platform {
  /** Rust target triple — the name cargo-dist uses for archives. */
  triple: string;
  /** Node `process.platform` value for this target. */
  nodePlatform: NodeJS.Platform;
  /** Node `process.arch` value for this target. */
  nodeArch: NodeJS.Architecture;
  /** napi library sub-package name (`@dirsql/lib-<slug>`). */
  libName: string;
  /** Wheel-style `os` constraint for the sub-package's package.json. */
  os: string[];
  /** Wheel-style `cpu` constraint for the sub-package's package.json. */
  cpu: string[];
  /** libc constraint (Linux only). */
  libc?: string[];
}

// JSON widens `nodePlatform` / `nodeArch` to `string`, which no data file can
// carry as `NodeJS.Platform` / `NodeJS.Architecture`. The cast is the only
// place the two vocabularies meet.
export const PLATFORMS = rows as readonly Platform[];

/** Node `${platform}-${arch}` → `@dirsql/lib-*` napi sub-package name. */
export function libTriples(): Record<string, string> {
  const out: Record<string, string> = {};
  for (const p of PLATFORMS) {
    out[`${p.nodePlatform}-${p.nodeArch}`] = p.libName;
  }
  return out;
}
